packages/flutter_tools/lib/src/asset.dart

```diff
@@ -0,0 +1,420 @@
+        return null;
+      }
+
+      List<_Asset> variants = <_Asset>[];
+      result[baseAsset] = variants;
+
+      // Find asset variants
+      String assetPath = baseAsset.assetFile.path;
+      String assetFilename = path.basename(assetPath);
+      Directory assetDir = new Directory(path.dirname(assetPath));
+
+      List<FileSystemEntity> files = assetDir.listSync(recursive: true);
+
+      for (FileSystemEntity entity in files) {
+        if (!FileSystemEntity.isFileSync(entity.path))
+          continue;
+
+        // Exclude any files in the given directories.
+        if (excludeDirs.any((String exclude) => entity.path.startsWith(exclude)))
+          continue;
+
+        if (path.basename(entity.path) == assetFilename && entity.path != assetPath) {
+          String key = path.relative(entity.path, from: baseAsset.base);
+          String assetEntry;
+          if (baseAsset.symbolicPrefix != null)
+            assetEntry = path.join(baseAsset.symbolicPrefix, key);
+          variants.add(new _Asset(base: baseAsset.base, assetEntry: assetEntry, relativePath: key));
+        }
+      }
+    }
+  }
+
+  // Add assets referenced in the fonts section of the manifest.
+  if (manifestDescriptor.containsKey('fonts')) {
+    for (Map<String, dynamic> family in manifestDescriptor['fonts']) {
+      List<Map<String, dynamic>> fonts = family['fonts'];
+      if (fonts == null) continue;
+
+      for (Map<String, dynamic> font in fonts) {
+        String asset = font['asset'];
+        if (asset == null) continue;
+
+        _Asset baseAsset = _resolveAsset(packageMap, assetBase, asset);
+        if (!baseAsset.assetFileExists) {
+          printError('Error: unable to locate asset entry in flutter.yaml: "$asset".');
+          return null;
+        }
+
+        result[baseAsset] = <_Asset>[];
+      }
+    }
+  }
+
+  return result;
+}
+
+_Asset _resolveAsset(
+  PackageMap packageMap,
+  String assetBase,
+  String asset
+) {
+  if (asset.startsWith('packages/') && !FileSystemEntity.isFileSync(path.join(assetBase, asset))) {
+    // Convert packages/flutter_gallery_assets/clouds-0.png to clouds-0.png.
+    String packageKey = asset.substring(9);
+    String relativeAsset = asset;
+
+    int index = packageKey.indexOf('/');
+    if (index != -1) {
+      relativeAsset = packageKey.substring(index + 1);
+      packageKey = packageKey.substring(0, index);
+    }
+
+    Uri uri = packageMap.map[packageKey];
+    if (uri != null && uri.scheme == 'file') {
+      File file = new File.fromUri(uri);
+      return new _Asset(base: file.path, assetEntry: asset, relativePath: relativeAsset);
+    }
+  }
+
+  return new _Asset(base: assetBase, relativePath: asset);
+}
+
+dynamic _loadFlutterYamlManifest(String manifestPath) {
+  if (manifestPath == null || !FileSystemEntity.isFileSync(manifestPath))
+    return null;
+  String manifestDescriptor = new File(manifestPath).readAsStringSync();
+  return loadYaml(manifestDescriptor);
```

Comment:
FWIW I tend to opt for `loadYamlNode` which is less dynamic and certain to return a YamlNode.

Since we're casting this result later maybe it'd be good to return a more specific type here and add an `is` check?

Dunno.  Sorry if this reads pedantic....